docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,871 @@
+
+export default {
+  environments: {
+    ssr: createWorkerdEnvironment({
+      build: {
+        outDir: '/dist/ssr',
+      },
+    }),
+    rsc: createWorkerdEnvironment({
+      build: {
+        outDir: '/dist/rsc',
+      },
+    }),
+  ],
+}
+```
+
+In this case we see how a Workerd environment can be set for both the default SSR environment and for a new custom RSC environment.
+
+## Plugins and environments
+
+### Accessing the current environment in hooks
+
+The Vite server has a shared plugin pipeline, but when a module is processed it is always done in the context of a given environment. The `environment` instance is available in the plugin context of `resolveId`, `load`, and `transform`.
+
+A plugin could use the `environment` instance to:
+
+- Only apply logic for certain environments.
+- Change the way they work depending on the configuration for the environment, which can be accessed using `environment.options`. The vite core resolve plugin modifies the way it resolves ids based on `environment.options.resolve.conditions` for example.
+
+```ts
+  transform(code, id) {
+    console.log(this.enviroment.options.resolve.conditions)
+  }
+```
+
+### Registering new environments using hooks
+
+Plugins can add new environments in the `config` hook:
+
+```ts
+  config(config: UserConfig) {
+    config.environments.rsc ??= {}
+  }
+```
+
+An empty object is enough to register the environment, default values from the root level environment config.
+
+### Configuring environment using hooks
+
+While the `config` hook is running, the complete list of environments isn't yet known and the environments can be affected by both the default values from the root level environment config or explicitely through the `config.environments` record.
+Plugins should set default values using the `config` hook. To configure each environment, they can use the new `configEnvironment` hook. This hook is called for each environment with its partially resolved config including resolution of final defaults.
+
+```ts
+  configEnvironment(name: string, options: EnvironmentOptions) {
+    if (name === 'rsc') {
+      options.resolve.conditions = // ...
+```
+
+### The `hotUpdate` hook
+
+- **Type:** `(ctx: HotContext) => Array<ModuleNode> | void | Promise<Array<ModuleNode> | void>`
+- **See also:** [HMR API](./api-hmr)
+
+The `hotUpdate` hook allows plugins to perform custom HMR update handling for a given environment. When a file changes, the HMR algorithm is run for each environment in series according to the order in `server.environments`, so the `hotUpdate` hook will be called multiple times. The hook receives a context object with the following signature:
+
+```ts
+interface HotContext {
+  file: string
+  timestamp: number
+  modules: Array<ModuleNode>
+  read: () => string | Promise<string>
+  server: ViteDevServer
+}
+```
+
+- `this.environment` is the module execution environment where a file update is currently being processed.
```

Comment:
The alternative to `this.environment` is having an `environment` param in every hook. We're using `this` because this is how vanilla Rollup already exposes the plugin context to hooks, and what we provide at the `environment` instance are Vite specific extensions to that context. See:
- https://github.com/vitejs/vite/pull/16261